mri/scripts/gridsearch.py

```diff
@@ -0,0 +1,289 @@
+        1 => View best results if present
+    """
+    if fourier_op is None and fourier_params is None:
+        raise ValueError('Both fourier_params and fourier_op cannot be None, '
+                         'please specify the fourier operator or a method to '
+                         'initialize it through fourier_params')
+    # Convert non-list elements to list so that we can create
+    # search space
+    init_classes = []
+    key_names = []
+    if linear_params is None:
+        linear_params = {
+            'init_class': WaveletN,
+            'kwargs':
+                {
+                    'wavelet_name': 'sym8',
+                    'nb_scale': 4,
+                }
+        }
+    if regularizer_params is None:
+        regularizer_params = {
+            'init_class': SparseThreshold,
+            'kwargs':
+                {
+                    'linear': Identity(),
+                    'weights': [0],
+                }
+        }
+    if optimizer_params is None:
+        optimizer_params = {
+            # Just following convention
+            'kwargs':
+                {
+                    'optimization_alg': 'fista',
+                    'num_iterations': 100,
+                }
+        }
+    for specific_params in [linear_params, regularizer_params,
+                            optimizer_params]:
+        for key, value in specific_params['kwargs'].items():
+            if not isinstance(value, (list, tuple, np.ndarray)):
+                specific_params['kwargs'][key] = [value]
+        # Obtain Initialization classes
+        if specific_params != optimizer_params:
+            init_classes.append(specific_params['init_class'])
+        # Obtain Key Names
+        key_names.append(list(specific_params['kwargs'].keys()))
+    # Create Search space
+    cross_product_list = list(itertools.product(
+        *linear_params['kwargs'].values(),
+        *regularizer_params['kwargs'].values(),
+        *optimizer_params['kwargs'].values(),
+    ))
+    test_cases = []
+    number_of_test_cases = len(cross_product_list)
+    if verbose > 0:
+        print('Total number of gridsearch cases : ' +
+              str(number_of_test_cases))
+    # Reshape data such that they match values for key_names
+    for i in range(number_of_test_cases):
+        # Iterator for i-th test case
+        iterator = iter(cross_product_list[i])
+        # Add the test case after reshaping the list
+        all_kwargs_values = []
+        for indivitual_param_names in key_names:
+            param_kwargs = {}
+            for key in indivitual_param_names:
+                param_kwargs[key] = next(iter(iterator))
+            all_kwargs_values.append(param_kwargs)
+        test_cases.append(_TestCase(
+            *init_classes,
+            *all_kwargs_values)
+        )
+
+    # Call for reconstruction
+    results = Parallel(n_jobs=n_jobs)(
+        delayed(test_cases[i].reconstruct_case)(
+            kspace_data=kspace_data,
```

Comment:
can't all this be part of test case? why pass it at run time?

Reply:
Techincally, we can move everything including fourier_op to test_case. 
I dont have any particular reason for this.

Only thing I am trying to do is to create reconstructor class in this function, as that results in iterations to get Lipchitz constant which I want to be parallel